s3/s3.go

```diff
@@ -1 +1,100 @@
 package s3
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/opensaucerer/bifrost/shared/errors"
+	"github.com/opensaucerer/bifrost/shared/types"
+)
+
+func (s SimpleStorageService) UploadFile(path, filename string) (types.UploadedFile, error) {
+	var ctx context.Context
+	var cancel context.CancelFunc
+	ctx = context.Background()
+	if s.DefaultTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(s.DefaultTimeout)*time.Second)
+		defer cancel()
+	}
+	// verify that file exists
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return types.UploadedFile{}, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+	// Create an AWS session
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(s.Region),
+		Credentials: credentials.NewStaticCredentials(s.AccessKey, s.SecretKey, "")},
```

Comment:
Currently, google has two ways of authenticating (service account file, and default credentials) and these two methods are implemented by the `newGoogleCloudStorage` function. 

S3, on the other hand, has about [4 methods](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html) and we need to implement all of them with the right order of priority like so...
- check if access and secret key are provided by the user and use
- else try to authenticate without any credentials. This means the session will try to find the credentials by itself using the 3 remaining methods. We should be sure to validate that a config was found using `_, err := sess.Config.Credentials.Get()` this is because s3 won't throw any auth error until when an API request is made and we don't want that.

Reply:
I switched to v2 of the sdk so [here](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/) is the doc for configuring it

Reply:
Awesome.
I think that work.